### ISSUE TITLE
Deno support

### ISSUE THREAD

**Issue by platon-ivanov** [rphlmr/react-router-hono-server#34]
Hello! This is a great and clean project! Is Deno v2 support on the timeline? What are the problems to be solved to make it possible?

**Comment by rphlmr** [comment]
Hello, thanks 🙏

I tried but my lack of knowledge was a problem. For some reason I have troubles to scope the deno language service/plugin to a file and my editor is full of red errors 😅

**Comment by NRS-DanielPartschev** [comment]
I am also looking forward for support of the Deno runtime. I've already made many attempts to handle react-router with hono inside Deno, but it hasn't quite worked out so far 😅

**Comment by jirizavadil** [comment]
@NRS-DanielPartschev here's my **deno** + **hono** + **react router v7 SSR** attempt

After reading two internets, mostly thanks to @rphlmr, @sergiodxa, @redabacha & @AlemTuzlak, i was able to piece this together. But it has completely separate prod & dev server, which is suboptimal. When `react-router-hono-server` supports Deno, i'll happily switch.

TLDR;
- `renderToReadableStream` instead of node's `renderToPipeableStream`
- `ssr.resolve.*` overrides in `vite.config.js`
- `Deno.serve(app.fetch)` in hono

---

```js
// vite.config.js

import deno from "@deno/vite-plugin"
import {reactRouter} from "@react-router/dev/vite"
import {defineConfig} from "vite"

export default defineConfig(() => {
  return {
    plugins: [deno(), reactRouter()],
    ssr: {
      resolve: {
        conditions: ["module", "deno", "node", "development|production"],
        externalConditions: ["deno", "node"],
      },
    },
  }
})
```

```js
// react-router.config.js

export default {
  ssr: true,
}
```

```jsonc
// deno.jsonc

{
  "tasks": {
    "dev": "react-router dev",
    "build": "react-router build",
    "start": "deno run --allow-net --allow-read --allow-env=HOST,PORT,NODE_ENV,IS_RR_BUILD_REQUEST ./server/main.js"
  },
  "nodeModulesDir": "auto",
  "imports": {
    "@deno/vite-plugin": "npm:@deno/vite-plugin@^1.0.4",
    "@react-router/dev": "npm:@react-router/dev@^7.5.3",
    "hono": "npm:hono@^4.7.8",
    "isbot": "npm:isbot@^5.1.27",
    "react": "npm:react@^19.1.0",
    "react-dom": "npm:react-dom@^19.1.0",
    "react-router": "npm:react-router@^7.5.3",
    "vite": "npm:vite@^6.3.5"
  }
}
```

```jsx
// src/app/entry.server.jsx

import {renderToReadableStream} from "react-dom/server"
import {ServerRouter} from "react-router"

export default async function handleRequest(request, responseStatusCode, responseHeaders, routerContext, loadContext) {
  // ...
  const body = await renderToReadableStream(<ServerRouter context={routerContext} url={request.url} />, {
    //...
  })
  // ...
  return new Response(body, {
    headers: responseHeaders,
    status: responseStatusCode,
  })
}
```

```js
// server/main.js

import {Hono} from "hono"
import {serveStatic} from "hono/deno"
import middlewareReactRouter from "./middleware/reactRouter.js"

const app = new Hono()

app.use("/assets/*", serveStatic({root: "./build/client"}))
app.use("*", middlewareReactRouter)

Deno.serve(app.fetch)
```

```js
// server/middleware/reactRouter.js

import * as build from "../../build/server/index.js"
import {createMiddleware} from "hono/factory"
import {createRequestHandler} from "react-router"

export default createMiddleware(async (c, next) => {
  const handler = createRequestHandler(build, "production")
  const loaderContext = {
    HELLO_FROM_HONO: "hello from hono",
  }
  return await handler(c.req.raw, loaderContext)
})
```

**Comment by jirizavadil** [comment]
Found a way how to use `react-router-hono-server` with Deno, however it's not pure Deno (eg. won't be able to deploy to Deno Deploy). Works in custom environment (eg. docker) where you can launch with custom command though.

To run this without any modifications outside Deno Deploy, this hack hacks:

```jsonc
// deno.jsonc

{
  "tasks": {
    "start": "deno run --allow-net --allow-read --allow-env=HOST,PORT,NODE_ENV,IS_RR_BUILD_REQUEST --unstable-node-globals ./build/server/index.js" // --unstable-node-globals is needed for react-router-hono-server, because it uses global instead of globalThis
  },
  "imports": {
    // overrides for react-router-hono-server's production build (Deno needs node prefixes)
    "fs": "node:fs",
    "crypto": "node:crypto",
    "stream": "node:stream",
    "http": "node:http",
    "http2": "node:http2"
  }
}
```

---

To make this pure Deno and be able to deploy to Deno Deploy, I believe that 'all we need'™️ from @rphlmr are these modifications to `build/server/index.js` (tested this manually and was able to deploy without errors):
- replace `global.*` with `globalThis.*`
- somehow get rid of the `serve()` at `:763` (Deno Deploy launches with `deno serve` which is a `Deno.serve(app.fetch)` wrapper for default export. the `serve()` causes launch of two listeners on two different ports.). Or replace default export with `Deno.serve(app.fetch)`, which would make it also work with `deno run`.

I tried to fork and deal with this, but can't wrap head around Vite plugins.

Ideally also replacing fs, crypto, stream, http, http2 with Deno's alternatives would be nice, but since Deno 2 not required.

Strangely, `process.env.NODE_ENV` seems to work fine, although it should also ideally be `Deno.env('NODE_ENV')`

Hope this could help with the implementation, @rphlmr

**Comment by rphlmr** [comment]
Released! https://github.com/rphlmr/react-router-hono-server/releases/tag/v2.26.0

Thanks @ktKongTong
